src/uwtools/rocoto.py

```diff
@@ -143,3 +65,249 @@ def validate_rocoto_xml(input_xml: OptionalPath) -> bool:
+        :param config: Configuration data for this element.
+        """
+        for name, coords in config.items():
+            for coord in coords:
+                SubElement(e, STR.cycledef, group=name).text = coord
+
+    def _add_workflow_log(self, e: Element, logfile: str) -> None:
+        """
+        Add <log> element(s) to the <workflow>.
+
+        :param e: The parent element to add the new element to.
+        :param logfile: The path to the log file.
+        """
+        SubElement(e, STR.log).text = logfile
+
+    def _add_workflow_tasks(self, e: Element, config: dict) -> None:
+        """
+        Add <task> and/or <metatask> element(s) to the <workflow>.
+
+        :param e: The parent element to add the new element to.
+        :param config: Configuration data for these elements.
+        """
+        for key, block in config.items():
+            tag, name = self._tag_name(key)
+            {STR.metatask: self._add_metatask, STR.task: self._add_task}[tag](e, block, name)
+
+    def _insert_doctype(self, xml: str) -> str:
+        """
+        Insert the <!DOCTYPE> block in a rendered XML document.
+
+        :param xml: The XML document rendered as a string.
+        :return: The XML document with the <!DOCTYPE> block inserted.
+        """
+        lines = xml.split("\n")
+        if doctype := self._doctype:
+            lines.insert(1, doctype)
+        return "\n".join(lines)
+
+    def _set_attrs(self, e: Element, config: dict) -> None:
+        """
+        Set attributes on an element.
+
+        :param e: The element to set the attributes on. :config: A config containing the attribute
+            definitions.
+        """
+        for attr, val in config[STR.attrs].items():
+            e.set(attr, str(val))
+
+    def _tag_name(self, key: str) -> Tuple[str, str]:
+        """
+        Split a metadata-bearing key into base tag and metadata.
+
+        :param key: A string of the form "tag_metadata" (or simply STR.tag).
+        :return: The base tag and metadata, separated.
+        """
+        m = re.match(r"^([^_]+)(_(.*))?$", key)
+        assert m  # validated config => regex match
+        tag, name = m[1], m[3]
+        return tag, name
+
+    def _tidy(self, e: Element) -> Element:
+        """
+        Construct a semantically-equivalent tree with child elements and attributes in sorted order.
+
+        :param e: The element tree to sort.
+        :return: A new element tree with child elements and attributes in sorted order.
+        """
+        # Create a new element named the same as the current element, add the attrs in sorted order,
+        # then add the (recursively tidied) children in sorted order.
+        tidy_e = Element(e.tag)
+        for attr, val in sorted(e.items(), key=lambda x: x[0]):
+            tidy_e.set(attr, val)
+        tidy_e.text = e.text
+        for child in sorted(list(e), key=lambda x: x.tag):
+            tidy_e.append(self._tidy(child))
+        return tidy_e
```

Comment:
We'll likely want to leave the output XML unsorted so that it comes out as defined by the user. Developers generally want to see tasks in the order they play out in the workflow where they can. This has historically been the only way to visualize easily what comes "next".

Reply:
Isn't order based on dependencies, rather than on order in the file? If tasks have no dependencies, aren't they executed concurrently?

Reply:
Obviously I declined to advocate further on this in this PR, but there's one thing I think I should mention: If we're operating under an assumption that the order of tasks or metatasks in UW YAML is somehow meaningful, will be preserved, etc., we should revisit that soon. YAML mappings are strictly **unordered**. Even if, by chance, they end up in the same order in Python today, key order in YAML is meaningless and should not be depended on (https://yaml.org/spec/1.2.2/#3221-mapping-key-order). If order matters, we should use a sequence. I'm not really sure that this is happening, though, so please ignore if not.